File: pipeline/src/pipeline/repository.py
"""DataRepository abstract interface (Protocol).

Frontend and any future consumer MUST read data only through this interface,
never directly from JSON file paths. This keeps storage swappable
(``JsonFileRepository`` now, ``SupabaseRepository`` later) with zero UI changes.

Interface only -- NO implementation in this file (see PROJECT_RULES / v2).
"""

from __future__ import annotations

from typing import Optional, Protocol, runtime_checkable

from .models import (
    DateIndex,
    Event,
    HealthSnapshot,
    PublishedData,
    SourcesState,
    Trend,
)


@runtime_checkable
class DataRepository(Protocol):
    """Unified read access to all published data."""

    def get_latest(self) -> Optional[PublishedData]:
        """Return the most recent PublishedData (by latest_date), or None."""
        ...

    def get_by_date(self, date: str) -> Optional[PublishedData]:
        """Return PublishedData for a specific YYYY-MM-DD, or None."""
        ...

    def get_history_index(self) -> DateIndex:
        """Return the global date index (index.json)."""
        ...

    def get_health(self) -> HealthSnapshot:
        """Return the data-source health snapshot (health.json)."""
        ...

    def get_sources_state(self) -> SourcesState:
        """Return runtime source states (sources_state.json)."""
        ...

    def get_by_category(self, category: str, date: Optional[str] = None) -> list[Trend]:
        """Return published trends of a category, optionally for a date."""
        ...

    def get_events(self, date: Optional[str] = None) -> list[Event]:
        """Return aggregated events, optionally for a date."""
        ...

File: frontend/lib/repository.ts
// DataRepository abstract interface (frontend side).
//
// The frontend MUST read published data only through this interface, never by
// importing JSON file paths directly. This keeps storage swappable
// (JsonFileRepository now, SupabaseRepository later) with zero UI changes.
//
// Interface only -- NO implementation in this file (see PROJECT_RULES / v2).

import type {
  DateIndex,
  Event,
  HealthSnapshot,
  PublishedData,
  SourcesState,
  Trend,
} from "./types";

export interface DataRepository {
  /** Return the most recent PublishedData (by latest_date), or null. */
  getLatest(): Promise<PublishedData | null>;
  /** Return PublishedData for a specific YYYY-MM-DD, or null. */
  getByDate(date: string): Promise<PublishedData | null>;
  /** Return the global date index (index.json). */
  getHistoryIndex(): Promise<DateIndex>;
  /** Return the data-source health snapshot (health.json). */
  getHealth(): Promise<HealthSnapshot>;
  /** Return runtime source states (sources_state.json). */
  getSourcesState(): Promise<SourcesState>;
  /** Return published trends of a category, optionally for a date. */
  getByCategory(category: string, date?: string): Promise<Trend[]>;
  /** Return aggregated events, optionally for a date. */
  getEvents(date?: string): Promise<Event[]>;
}
